### ISSUE TITLE
Cleaning up the issue list

### ISSUE THREAD

**Issue by RoiEXLab** [triplea-game/triplea#11163]
So there are a lot of issues created that don't add any value. I'd like to discuss which types we can close without any consequences and which ones should be kept because they are an actual issue that can be tackled.

## No-Brainers (in my opinion)

1. OutOfMemoryError - there's not much we can do about them unless it's something where 10GB of RAM are not sufficient and there's a memory leak somewhere
2. MISSING UNIT IMAGE & missing unit icon - Indicates a problem with a map, not with the engine. IIRC newer versions of the engine don't create a ticket for this any longer.
3. ? Any other suggestions?

WDYT @DanVanAtta ? Anything that should be added to the list? Any additional thoughts to remove the repeating issues that clutter the list?

**Comment by RoiEXLab** [comment]
Also there are several issues that are not fixable for us, but would fall into a category "needs better error message" to communicate to the user that we're not the problem, the users system is.
Examples: #11069 and #8702 and probably a lot more

Should we introduce an extra label for those?

**Comment by beelee1** [comment]
yea i think missing image is fixed by adding a generic image for 2.6 

I try and go through and wipe out duplicates when I can, but that only works so well and not very efficient.

**Comment by DanVanAtta** [comment]
We typically have closed error reports once a new version is out. Mostly so
ws can focus on known open problems and there is still plenty to work from
on the new problems.

I think 2.5 error reports can be closed in bulk at this juncture (making
this easier was a reason to add a version labels to error reports)

I do think we are lacking frequency information on error reports. There is
a de-duping mechanism so we are really not seeing the top hitters.

### For bad error messages

I would really encourage is to think deeply about these. There us a lot of
code that makes naive assumptions and declares perfectly normal situations
errors. I think a lot of these are bugs if omission. Eg: not enough space
in disk, network disconnected. The error reporting functionality has an
example where if the lobby is not available then the 'report error' button
is not shown .

Failing this, for true errors, the error message should be useful and
directed to users to tell them what went wrong and how they can recover.
Users often restart the game anytime they see any error, and sometimes the
code throws errors as control flow rather than legit reason (eg:
GameoverException)

To rub it in a bit, each error report generated from uncaught exception is
a fully halted, possibly unrecoverable game. Errors from logger statements
are a 50/50 bad logging and lack of error checking and handling. That is
all to say, some of these errors are player losing events (if you spend 6
hours on a game to then lose it, it gives a lot of pause before starting
another).

While a number of things we cannot fix, it still behooves us to present
good messages to the user and not just a dead game with
"nullpointeexception" as the message to the user.

**Comment by RoiEXLab** [comment]
@DanVanAtta You're raising good points. Especially the `GameOverException` is what I found to be very confusing in the past, mainly because it is unclear where this exception is thrown in the first place, but also because we don't have a unified way to catch and handle the exception not to mention this approach does not work well in combination with multi threading.

On a more general note: I still think we need a nicer solution for error tracking. There are currently 646 open issues, where most of them are probably obsolete, either because the bug has been fixed because of some refactoring or because there exist like 6 duplicates of the same NPE that hasn't been fixed since it first occurred. This is especially tragic because I think there is a lot of low hanging fruit there that can be fixed with a couple of lines of code (see #11164 and #11165 for example) and there's no way to figure out how many of those open issues fall into this category and are worth investing like half an hour on the weekend.

**Comment by DanVanAtta** [comment]
I thing I had failed to mention, there are a lot of places where we should instead log a warning (which prompts user but does not give an option to send an error report) instead of an error. In combination with this, we can update the error message, eg: "It appears your internet connection has been interrupted. Please check your connectivity and try again". The warning message also automatically adds "if this keeps happening, contact us".

I'm interested to learn more about your concerns @RoiEXLab regarding a nicer solution. I still think frequency is maybe one of the biggest missing items.

 > either because the bug has been fixed

The server does not allow for duplicate errors to be reported. If an error has already been reported, the issue link is given to the user instead of the option to submit a new report. This is most problematic where error messages are dynamic. I think this caused a lot of noise for the 'missing unit images' since each one is almost unique, but my impression is otherwise that most errors (other than the missing images) typically do not have such dynamic text in them. So, if a bug is fixed, and if the error report has a stable error message - we can close the issue and it does not come up again.

> there's no way to figure out how many of those open issues fall into this category and are worth investing like half an hour on the weekend.

I think this is further reason to just close all of the 2.5 issues. The fact it's been such a long time since last release makes this more of a problem. Anything cut against 2.6 should be a current issue.

> I think there is a lot of low hanging fruit 

I agree, a large number of errors likely can be fixed without too much effort.

**Comment by beelee1** [comment]
_I think there is a lot of low hanging fruit_

maybe these would be good to turn the ISU kids lose on

**Comment by DanVanAtta** [comment]
@beelee1 already done! They picked up a 2.6 error report issue

**Comment by DanVanAtta** [comment]
It does seem like the current topic of conversation is this:
> I still think we need a nicer solution for error tracking. 

@RoiEXLab  do you think the problem tracker could help fit in as a part of solution to helping making it easier to keep track of issues? What would you say are the biggest problems we have currently?

From what I've researched about bug tracking software (FWIW), the requirements of bug tracking is generally for any given defect, to be able to ascertain the following 5 items:
- state
- assignment
- priority
- description & problem details, repro steps
- mandate for specific corrective action

Currently with how we mange issues, I see some possible improvement for each area.

**Comment by DanVanAtta** [comment]
@beelee1 , @panther2, @RoiEXLab , while we are managing issues.. I could use some help.

We need to go through this list: [(issues labeled with error report and not labeled with 2.6](https://github.com/triplea-game/triplea/issues?page=4&q=is%3Aopen+is%3Aissue+label%3A%22Error+Report%22++-label%3A%222.6%22+-label%3A%222.5%22), and label any of the '2.6' issues with the '2.6' label. That would make it a clean list of non-2.6 error report issues which we could then close entirely.

**Comment by beelee1** [comment]
@DanVanAtta ok so the entire current issues list. The oldest one is #6375 may 3 2020. Does 2.6 predate this or is it later ? We could obviously close any depending on 2.6 date.

Would there be any preference to go by oldest or recently updated etc ... ? If it doesn't reference 2.6, just close with comment "2.5 closing" or some such ?

**Comment by DanVanAtta** [comment]
@beelee1 The link I posted, is filtered list with about 15 pages. We just need to fish out the ones with 2.6 in their title and label them with 2.6; This can be done in bulk without opening each individual one.

I figure once we've done that, and there no more 2.6 error reports in that list, anything remaining is 2.5 and below; IMO we would just close them in bulk without comment.

The release notes indicate 2.5 launched in Nov 2020, so there is some overlap. Non-error report issues I think we need to probably address individually (not in bulk), even if they are older.

**Comment by DanVanAtta** [comment]
FWIW, I'm thinking the '2.6' and '2.5' (etc..) labels are to help sort the 'error report' labels. I was _not_ thinking of using them to indicate the minimum version of the game where a problem was discovered (though, I am not opposed to the idea, just it was not my initial intent). The latter, using a version label to indicate min-impacted-version, could be useful - but would be harder to upkeep. I also suspect once we have the 'error report' list pruned down to just '2.6', it'll be less important to label issues with game version. I'm interested if there are non-error-report use-cases where labeling game version is of notable help.

**Comment by DanVanAtta** [comment]
@RoiEXLab , I've given you a lot to read.. 
@beelee1  / @RoiEXLab , I feel like we could benefit perhaps from a "traiged" label or something. I've left a few comments in some issues with what we need to do and/or check and have left them open. Some of these issues have received the "good first issue" label because the task is clear and does not need much background. Anyways, a label to indicate "the task is clear" would perhaps be of benefit to indicate issues we have looked at, confirmed, and know what we need to do. Any thoughts if that would be really helpful (or just upkeep?), and any ideas for what to name that label?

**Comment by beelee1** [comment]
ok so if it doesn't say 2.6 in the title without opening it ignore it and if it does add the 2.6 label ? This won't miss any that way ? I opened the oldest one which is 2.0.1 but it didn't say that in the title

**Comment by beelee1** [comment]
yea don't want duplicate work. I'm showing 336 issues in the filtered list. might be best for me just to go through and close anything that doesn't say 2.6 ? i don't think i know enough to sort otherwise. Other than just adding the 2.6 label when i see it in the title

**Comment by DanVanAtta** [comment]
> Other than just adding the 2.6 label when i see it in the title

You got it :+1: 

> might be best for me just to go through and close anything that doesn't say 2.6 ?

You can do this! Feel free to go ahead if that is convenient and most efficient. You might find this just ever so slightly easier to do once all of the 2.6 issues are labelled and filtered out.

**Comment by beelee1** [comment]
right on I'll peck away at it starting with oldest. I always go the conservative route when doing stuff when i'm not sure. Don't wanna create more work lol Basically i just add rogers games and anyone else who might add one :)

**Comment by beelee1** [comment]
@DanVanAtta got it knocked down to 44. The ones that are left are mostly 2.6 and maybe a dozen i wasn't sure of

**Comment by DanVanAtta** [comment]
Good stuff @beelee1  - thank you  :+1:

**Comment by Cernelius** [comment]
> 1. OutOfMemoryError - there's not much we can do about them unless it's something where 10GB of RAM are not sufficient and there's a memory leak somewhere

You can do https://github.com/triplea-game/triplea/issues/9654

Or even just set it at 4G. If one has not enough, he or she is probably on a 32 bit and that would remain unchanged.

**Comment by asvitkine** [comment]
Just FYI: I went out and closed nearly all 2.5 issues. I left some open that had some discussion that may still be relevant.
